pkg/database/realm.go

```diff
@@ -513,3 +520,51 @@ func (r *Realm) CreateSigningKeyVersion(ctx context.Context, db *Database) (stri
 
 	return signingKey.GetKID(), nil
 }
+
+// DestroySigningKeyVersion destroys the given key version in both the database
+// and the key manager. ID is the primary key ID from the database. If the id
+// does not exist, it does nothing.
+func (r *Realm) DestroySigningKeyVersion(ctx context.Context, db *Database, id interface{}) error {
+	manager := db.signingKeyManager
+	if manager == nil {
+		return ErrNoSigningKeyManager
+	}
+
+	if err := db.db.Transaction(func(tx *gorm.DB) error {
+		// Load the signing key to ensure it actually exists.
+		var signingKey SigningKey
+		if err := tx.
+			Set("gorm:query_option", "FOR UPDATE").
+			Table("signing_keys").
+			Where("id = ?", id).
+			Where("realm_id = ?", r.ID).
+			First(&signingKey).
+			Error; err != nil {
+			if IsNotFound(err) {
+				return nil
+			}
+			return fmt.Errorf("failed to load signing key: %w", err)
+		}
+
+		if signingKey.Active {
+			return fmt.Errorf("cannot destroy active signing key")
+		}
+
+		// Delete the signing key from the key manager - we want to do this in the
+		// transaction so, if it fails, we can rollback and try again.
+		if err := manager.DestroyKeyVersion(ctx, signingKey.KeyID); err != nil {
+			return fmt.Errorf("failed to destroy signing key in key manager: %w", err)
+		}
+
+		// Successfully deleted from the key manager, now remove the record.
+		if err := tx.Delete(&signingKey).Error; err != nil {
+			return fmt.Errorf("failed to delete signing key from database: %w", err)
```

Comment:
but... key material was destroyed. I would be sure to add that to the error message.

Reply:
Done